### PR TITLE
Control optical sensor power saving modes

### PR DESCRIPTION
**Pull request by Qbicz**
Added:
- Kconfig options to choose when the optical sensor downshifts to power saving modes
- configuration channel events to control the downshift times at runtime
- updated Python application to support that

There is also a small cleanup in the configuration which is unrelated.

### CODE REVIEW

**Comment by Qbicz** [comment]
@pdunaj can you review? @jhn-nordic could you also take a look?

File: samples/nrf_desktop/src/hw_interface/Kconfig [hunk 1]
```diff
@@ -63,6 +63,39 @@ config DESKTOP_OPTICAL_DEFAULT_CPI
...
 	  using the configuration channel.
 
+config DESKTOP_OPTICAL_RUN_DOWNSHIFT_TIME_MS
+	int "Default run downshift time of the mouse optical sensor"
+	depends on DESKTOP_MOTION_OPTICAL_ENABLE
+	default 500
+	range 10 2550
+	help
+	  Time after which sensor goes from "Run" to "Rest 1" mode.
+
```
Comment (pdunaj) [inline samples/nrf_desktop/src/hw_interface/Kconfig]: excessive blank line
also in other options

File: samples/nrf_desktop/src/hw_interface/motion_optical.c [hunk 2]
```diff
@@ -436,6 +439,58 @@ static void update_cpi(u16_t cpi)
...
+{
+	/* Set downshift time:
+	 * - Run downshift time (from Run to Rest1 mode)
+	 * - Rest 1 downshift time (from Rest1 to Rest2 mode)
+	 * - Rest 2 downshift time (from Rest2 to Rest3 mode)
+	 */
+	u32_t maxtime;
+	u32_t mintime;
+
+	if (reg_addr == OPTICAL_REG_RUN_DOWNSHIFT) {
```
Comment (pdunaj) [inline samples/nrf_desktop/src/hw_interface/motion_optical.c]: for this kind of flow control please use `switch/case`

File: samples/nrf_desktop/src/hw_interface/motion_optical.c [hunk 3]
```diff
@@ -436,6 +439,58 @@ static void update_cpi(u16_t cpi)
...
+		return;
+	}
+
+	if ((time > maxtime) || (time < mintime)) {
+		LOG_WRN("Downshift time %u out of range", time);
+		return;
+	}
+
+	/* Convert time to register value */
+	u8_t value = time / mintime;
```
Comment (pdunaj) [inline samples/nrf_desktop/src/hw_interface/motion_optical.c]: value will never be zero? is zero not a correct option?

File: samples/nrf_desktop/src/hw_interface/motion_optical.c [hunk 4]
```diff
@@ -744,13 +810,47 @@ static void optical_thread_fn(void)
 				cpi = new_cpi;
 				update_cpi(cpi);
 			}
+
```
Comment (pdunaj) [inline samples/nrf_desktop/src/hw_interface/motion_optical.c]: please hide this code behind some function

File: samples/nrf_desktop/src/hw_interface/motion_optical.c [hunk 5]
```diff
@@ -631,6 +686,14 @@ static int init(u16_t cpi)
 		goto error;
 	}
 
+	/* Set initial power mode downshift times. */
+	update_downshift_time(OPTICAL_REG_RUN_DOWNSHIFT,
+			      CONFIG_DESKTOP_OPTICAL_RUN_DOWNSHIFT_TIME_MS);
```
Comment (pdunaj) [inline samples/nrf_desktop/src/hw_interface/motion_optical.c]: I think we should pass it as an argument from the caller (just like with cpi)
this way we set it to default value in one place only

Comment (pdunaj) [reply samples/nrf_desktop/src/hw_interface/motion_optical.c]: (also for other values)

File: samples/nrf_desktop/src/hw_interface/motion_optical.c [hunk 6]
```diff
@@ -744,13 +810,47 @@ static void optical_thread_fn(void)
...
 	}
 	/* This thread is supposed to run forever. */
 	__ASSERT_NO_MSG(false);
 }
 
+static void set_downshift_time_from_bytes(atomic_t *sensor_val,
+					  const u8_t *data)
+{
+	/* Convert little endian bytes to an unsigned integer. */
+	u32_t new_time = data[0] + (data[1] << 8)
```
Comment (pdunaj) [inline samples/nrf_desktop/src/hw_interface/motion_optical.c]: I think we have utility function that does the conversion for us (check hids)

File: samples/nrf_desktop/src/hw_interface/motion_optical.c [hunk 7]
```diff
@@ -847,9 +947,37 @@ static bool event_handler(const struct event_header *eh)
...
 			const struct config_event *event = cast_config_event(eh);
 
-			if (event->id == CONFIG_EVENT_ID_MOUSE_CPI) {
-				u16_t new_cpi = event->data[0] + (event->data[1] << 8);
-				atomic_set(&sensor_cpi, new_cpi);
+			switch (event->id) {
+			case CONFIG_EVENT_ID_MOUSE_CPI:
+				atomic_set(&sensor_cpi, (event->data[0] +
+					(event->data[1] << 8)));
+				k_sem_give(&sem);
```
Comment (pdunaj) [inline samples/nrf_desktop/src/hw_interface/motion_optical.c]: please don't semgive if we are fetching (it may break the sync as we will get additional sample)
we need to fix it is some smarter way but let's do it in another PR

File: samples/nrf_desktop/src/hw_interface/motion_optical.c [hunk 8]
```diff
@@ -847,9 +947,37 @@ static bool event_handler(const struct event_header *eh)
 		if (is_config_event(eh)) {
 			const struct config_event *event = cast_config_event(eh);
 
-			if (event->id == CONFIG_EVENT_ID_MOUSE_CPI) {
-				u16_t new_cpi = event->data[0] + (event->data[1] << 8);
-				atomic_set(&sensor_cpi, new_cpi);
+			switch (event->id) {
```
Comment (pdunaj) [inline samples/nrf_desktop/src/hw_interface/motion_optical.c]: please hide this code behind a function

File: samples/nrf_desktop/src/hw_interface/motion_optical.c (hunk 3 above)
Comment (Qbicz) [reply samples/nrf_desktop/src/hw_interface/motion_optical.c]: Zero is not a correct option, at least for this sensor. Added assert, I believe assert check during development is enough

File: samples/nrf_desktop/src/hw_interface/motion_optical.c (hunk 5 above)
Comment (Qbicz) [reply samples/nrf_desktop/src/hw_interface/motion_optical.c]: done

File: samples/nrf_desktop/src/hw_interface/motion_optical.c (hunk 2 above)
Comment (Qbicz) [reply samples/nrf_desktop/src/hw_interface/motion_optical.c]: done

File: samples/nrf_desktop/src/hw_interface/Kconfig (hunk 1 above)
Comment (Qbicz) [reply samples/nrf_desktop/src/hw_interface/Kconfig]: done

File: samples/nrf_desktop/src/hw_interface/motion_optical.c (hunk 6 above)
Comment (Qbicz) [reply samples/nrf_desktop/src/hw_interface/motion_optical.c]: `sys_get_le16()` and `sys_get_le32()` look useful, thanks!

File: samples/nrf_desktop/src/hw_interface/motion_optical.c (hunk 8 above)
Comment (Qbicz) [reply samples/nrf_desktop/src/hw_interface/motion_optical.c]: done

File: samples/nrf_desktop/src/hw_interface/motion_optical.c (hunk 4 above)
Comment (Qbicz) [reply samples/nrf_desktop/src/hw_interface/motion_optical.c]: done

File: samples/nrf_desktop/src/hw_interface/motion_optical.c (hunk 7 above)
Comment (Qbicz) [reply samples/nrf_desktop/src/hw_interface/motion_optical.c]: Added a simple state check. It works, sync is not broken

**Comment by Qbicz** [comment]
@pdunaj, can you revisit?

File: samples/nrf_desktop/src/hw_interface/motion_optical.c [hunk 9]
```diff
@@ -625,12 +697,20 @@ static int init(u16_t cpi)
 	}
 
 	/* Set initial CPI resolution. */
-	update_cpi(cpi);
-	atomic_set(&sensor_cpi, cpi);
+	update_cpi(options->cpi);
+	atomic_set(&sensor_cpi, options->cpi);
 	if (err) {
```
[the remark this answers is not in the document]
Comment (Qbicz) [reply samples/nrf_desktop/src/hw_interface/motion_optical.c]: removed

**Comment by Qbicz** [comment]
Hi @pizi-nordic , can you merge?
